### DISCUSSION TITLE
Monkeytype User Quotes

### DISCUSSION THREAD

**Discussion by tajhans** [monkeytypegame/monkeytype#1801] [Ideas]
I was thinking maybe there could be a way that users could add quotes straight from the site. Maybe there would be a form. Select what length the quote is, add the quote, and source it. Of course this is a way for trolls to reek havoc, so there could be a report feature. Still, this may not be viable because of the trolls but I do think it would be better than going to the GitHub repo to add quotes.

**Comment by lukew3** [comment]
I agree that users should be able to submit quotes through the site, especially considering that uploading quotes via discord seems to be deprecated. Maybe there could be a quote moderator web page where moderators can approve or deny quotes. Could also ban users from suggesting quotes if moderators so choose.

**Reply by unrealapex** [reply]
I also agree, adding quotes directly would be very easy. We could do it by the honor system, but there's always one person that ruins it for everyone. Some sort of verification would help prevent abuse.

**Reply by tajhans** [reply]
I think this is great. @UnrealApex I do think it is important to have some verification as you mentioned that one person.

**Reply by unrealapex** [reply]
Most of our users are trustworthy, so I guess having multiple users(that we pick) to approve submissions.

**Reply by tajhans** [reply]
A whitelist is an interesting idea. Maybe people could apply too.

**Reply by unrealapex** [reply]
That sounds good, or have one of the contributors or people on the discord server review them.

**Reply by tajhans** [reply]
Yeah, like mods.

**Reply by unrealapex** [reply]
Indeed.

**Reply by lukew3** [reply]
@Miodec I should have time to work on this tomorrow. Do you approve? I was kind of thinking about moving quotes to a mongo collection, but I don't have a problem with leaving it as git either. Thoughts on this?

**Reply by Miodec** [reply]
I was thinking about building a separate site, something like quotes.monkeytype.com where users can submit quotes, vote to accept new ones, and moderators and accept quotes.

@lukew3 whats the advantage of moving quotes to mongo?

**Reply by lukew3** [reply]
@Miodec Maybe, but I don't know if a separate site is necessary. I like the idea of having a box pop-up after clicking on an add quote button from the test page. Something like this plus button:
![Screenshot from 2021-09-09 10-09-50](https://user-images.githubusercontent.com/47042841/132701643-5df7bb2d-3efb-4373-8d06-76434cd630fc.png)
I don't see a point in having users vote for new quotes, as there shouldn't be too many for mods to handle. I'm thinking that maybe there should just be a seperate static html page with js where mods can handle quotes. You could use current authentication to allow mods to use their regular monkeytype account to approve changes, instead of making a seperate mod login.

Moving quotes to mongo would help manage the size of the repo and might make managing quotes easier. Right now quotes only take up 3.38 MB, whereas languages take up 25MB and the actual source code is about 2 MB. Doesn't seem to be a big issue at the moment, I'd say we just stick with using git.

**Reply by tajhans** [reply]
I agree with @lukew3. However, the files might add up after a while. It could also be easier to manage.

**Reply by Miodec** [reply]
Or to reduce the clutter in the top right (I keep thinking I should redesign that part of the page), put the button in the quote search menu 
![image](https://user-images.githubusercontent.com/13181393/132720459-48a01aa8-76e9-4b2f-9a61-9e9403b108a1.png)

I can build the admin/quote verifying panel, if you want to work on the submitting the quotes and storing them in a 'waiting for review' database collection.

If we wanna worry about the repo size then I think limiting word lists to 100k would be the first step. English 450k is full of gibberish anyway.

**Reply by tajhans** [reply]
@Miodec Yeah, I think that is a great idea. I feel like that makes sense to put it there. I don't think repo size is that big of a deal. I just mentioned it because it could be a possibility. Also would review be something built into the site or would they be checked in the database and then moved to the actual collection?

**Reply by unrealapex** [reply]
> I agree with @lukew3. However, the files might add up after a while. It could also be easier to manage.

It isn't likely though. Even if they did increase to a large size, we could always minify if the code hasn't been minified yet.
